ERCS/erc-7813.md

```diff
@@ -0,0 +1,498 @@
+---
+eip: 7813
+title: Store, Table-Based Introspectable Storage
+description: On-chain tables for automatic indexing and introspectable state
+author: alvarius (@alvrs), dk1a (@dk1a), frolic (@holic), ludens (@ludns), vdrg (@vdrg), yonada <yonada@proton.me>
+discussions-to: https://ethereum-magicians.org/t/erc-7813-store-table-based-introspectable-storage/21628
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-11-08
+---
+
+## Abstract
+
+This standard introduces a flexible on-chain storage pattern that organizes data into structured tables with schemas, similar to a traditional database. This approach allows new tables to be added at runtime without impacting existing contracts, thereby simplifying upgrades and extensions. By providing a unified interface for data access, the standard enables any contract or off-chain service to read stored data without the need for custom getter functions. Additionally, by standardizing event emissions for state changes it enables automatic, schema-aware indexing.
+
+## Motivation
+
+The absence of consistent standards for on-chain data management in smart contracts can lead to rigid implementations, tightly coupled contract logic with off-chain services, and challenges in updating or extending a contract’s data layout without breaking existing integrations.
+
+Using the storage mechanism defined in this ERC provides the following benefits:
+
+1. **Automatic Indexing**: By emitting consistent, standardized events during state changes, off-chain services can automatically track on-chain state and provide schema-aware indexer APIs.
+2. **Elimination of Custom Getter Functions**: Any contract or off-chain service can read stored data through a consistent interface, decoupling smart contract implementation from specific data access patterns and reducing development overhead.
+3. **Simpler Upgradability**: This pattern leverages unstructured storage, making it easier to upgrade contract logic without the risks associated with using a fixed storage layout.
+4. **Flexible Data Extensions**: New tables can be added at runtime without without breaking existing integrations with other data consumers.
+5. **Reduced gas costs**: Using efficient data packing reduces gas costs for both storage and event emissions.
+
+## Specification
+
+### Definitions
+
+#### Store
+
+A smart contract that implements the interface proposed by this ERC and organizes data in Tables. It emits events for each data operation so that off-chain components can replicate the state of all tables.
+
+#### Table
+
+A storage structure that holds **Records** sharing the same **Schema**.
+
+- **On-chain Table**: Stores its state on-chain and emits events for off-chain- indexers.
+- **Off-chain Table**: Does not store state on-chain but emits events for off-chain indexers.
+
+#### Record
+
+A piece of data stored in a **Table**, addressed by one or more keys.
+
+#### ResourceId
```

Comment:
If these headings are literals from your code (and it looks like they might be), you should enclose them in backticks (\`). That goes for any code snippets in the text as well.